### PR TITLE
Update inst-seg model export to include feature vector and saliency map

### PR DESCRIPTION
**Pull request by eugene123tw**
### Summary

This PR fixed explain model for MaskRCNN/RTMDet-Inst/MaskRCNN-TV instance segmentation model
- Export feature vector when `explain_mode` is ON
- Export dummy saliency map when `explain_mode` is ON (saliency map is handled in MAPI)
- Add MaskRCNN-TV explainable feature for PyTorch model

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [x] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Review by sovrasov — DISMISSED** [review]

**Review by sovrasov — DISMISSED** [review]

**Comment by sovrasov** [comment]
@eugene123tw thanks for the prompt fix! To address the issue in Geti 2.5 the changes should target OTX 2.2 release branch as well

**Review by chuneuny-emily — DISMISSED** [review]

**Comment by harimkang** [comment]
i create PR that include same changes to 2.2 (https://github.com/openvinotoolkit/training_extensions/pull/4056)

File: src/otx/core/model/instance_segmentation.py [hunk 1]
```diff
@@ -482,8 +484,8 @@ def _forward_explain_inst_seg(
             # Export case, consists of tensors
             # For OV task saliency map are generated on MAPI side
             saliency_map = torch.empty(1, dtype=torch.uint8)
-        elif isinstance(predictions, list) and isinstance(predictions[0], InstanceData):
-            # Predict case, consists of InstanceData
+        elif isinstance(predictions, list) and isinstance(predictions[0], (InstanceData, dict)):
+            # Predict case, consists of InstanceData or dict
```
Comment (sovrasov) [inline src/otx/core/model/instance_segmentation.py:488]: Is `self.explain_fn` on the next line ready for a dict?

Comment (eugene123tw) [reply src/otx/core/model/instance_segmentation.py:488]: yeah, `self.explain_fn = InstSegExplainAlgo.func`  supports list of dictionaries.

```py
def func(
        self,
        predictions: list[InstanceData] | list[dict[str, Tensor]],
        _: int = -1,
    ) -> list[np.array]:
        """Generate saliency maps from predicted masks by averaging and normalizing them per-class.

        Args:
            predictions (list[InstanceData] | list[dict[str, Tensor]): Predictions of Instance Segmentation model.

        Returns:
            torch.Tensor: Class-wise Saliency Maps. One saliency map per each class - [batch, class_id, H, W]
        """
        batch_saliency_maps = []
        for prediction in predictions:
            class_averaged_masks = self.average_and_normalize(prediction, self.num_classes)
            batch_saliency_maps.append(class_averaged_masks)
        return torch.stack(batch_saliency_maps)

@classmethod
    def average_and_normalize(
        cls,
        pred: InstanceData | dict[str, torch.Tensor],
        num_classes: int,
    ) -> np.array:
        if isinstance(pred, dict):
            masks, scores, labels = pred["masks"], pred["scores"], pred["labels"]
```

**Review by eunwoosh — DISMISSED** [review]

**Comment by sungchul1** [comment]
@eugene123tw could you describe the purpose and changes of this PR in Summary?

File: src/otx/algo/instance_segmentation/segmentors/two_stage.py [hunk 2]
```diff
@@ -226,19 +226,44 @@ def export(
...
+            batch_img_metas (list[dict]): image information.
+            explain_mode (bool): whether to return feature vector.
 
+        Returns:
+            tuple[torch.Tensor, torch.Tensor, torch.Tensor] | dict:
+                - bboxes (torch.Tensor): bounding boxes.
+                - labels (torch.Tensor): labels.
+                - masks (torch.Tensor): masks.
+                - feature_vector (torch.Tensor, optional): feature vector.
+                - saliency_map (torch.Tensor, optional): saliency map.
```
Comment (sungchul1) [inline src/otx/algo/instance_segmentation/segmentors/two_stage.py]: Could you add an additional description of these properties being about dict output?

Comment (eugene123tw) [reply src/otx/algo/instance_segmentation/segmentors/two_stage.py]: Updated.

**Review by sungchul1 — APPROVED** [review]

**Review by eunwoosh — APPROVED** [review]

**Pull request merged**